mhw_analysis/systems/src/build.c

```diff
@@ -89,14 +105,54 @@ void first_pass(char *cube, int *mask, int *shape, int *parent, int *category) {
                 if (cube[idx] == 0)
                     continue;
 
-                // prior_labels = mask[i-1:i+2,j-1:j+2,k-1:k+2].flatten()
-                // Fill prior_labels and check for non-zero values
+                // Search for neighboring pixels
                 // Also determine the minimum value of the label
                 count = 0;
                 all_zero = 0;
                 minlabel = maxnlabels;
-                for (ii=i-1; ii<=i+1; ii++) // Latitude
-                    for (jj=j-1; jj<=j+1; jj++) // Longitude
+
+                // Reset
+                jp=1;
+                jm=1;
+                ip=1;
+                im=1;
+
+                // Indian Ocean at 145E; E/W boundary
+                if (j==indian_lon0 && i<=indian_lat0)
+                    jp = 0;
+                if (j==indian_lon0+1 && i<=indian_lat0)
+                    jm = 0;
+                // Indian Ocean at 100E; E/W boundary
+                if (j==indian_lon1 && i>=indian_lat1a-1 && i<=indian_lat1b)
+                    jp = 0;
+                if (j==indian_lon1+1 && i>=indian_lat1a && i<=indian_lat1b)
+                    jm = 0;
+                // Indian Ocean 0deg N/S boundary
+                if (i==indian_lat1a && j>=indian_lon1 && j<=indian_lon0+1)
+                    ip = 0;
+                if (i==indian_lat1a+1 && j>=indian_lon1 && j<=indian_lon0+1)
+                    im = 0;
+                // Indian Ocean at 20E; E/W boundary
+                if (j==indian_lon2 && i<=indian_lat2)
+                    jp = 0;
+                if (j==indian_lon2+1 && i<=indian_lat2)
+                    jm = 0;
+
+                // Pacific Ocean at 70W; E/W boundary
+                if (j==pacific_lon0 && i<=pacific_lat0)
+                    jp = 0;
+                if (j==pacific_lon0+1 && i<=pacific_lat0)
+                    jm = 0;
+
+                // Pacific Ocean at 66N; N/S boundary
+                if (i==pacific_lat1 && j>=pacific_lon1a && j<=pacific_lon1b)
+                    ip = 0;
+                if (i==pacific_lat1+1 && j>=pacific_lon1a && j<=pacific_lon1b)
+                    im = 0;
+
+                // Do it
+                for (ii=i-im; ii<=i+ip; ii++) // Latitude
+                    for (jj=j-jm; jj<=j+jp; jj++) // Longitude
                         for (kk=k-1; kk<=k+1; kk++) {  // Time
```

Comment:
5 FOR LOOPS 💀 💀 I assume there's no easy way to make this faster, but if it's a priority it's something to be looked into

Reply:
Not with C